src/mtcp/mtcp_restart.c

```diff
@@ -668,6 +668,37 @@ static void unmap_memory_areas_and_restore_vdso(RestoreInfo *rinfo)
     }
     MTCP_ASSERT(vdso == rinfo->vdsoStart);
   }
+#if defined(__i386__)
+    // In commit dec2c26c1eb13eb1c12edfdc9e8e811e4cc0e3c2 , the mremap
+    //   code above was added, and then caused a segfault on restart for
+    //   __i386__ in CentOS 7.  In that case ENABLE_VDSO_CHECK was not defined.
+    //   This version was needed in that commit for __i386__
+    //   (i.e., for multi-arch.sh) to succeed.
+    // Newer Linux kernels, such as __x86_64__, provide a separate vsyscall segment
+    //   for kernel calls while using vdso for system calls that can be
+    //   executed purely in user space through kernel-specific user-space code.
+    // On older kernels like __x86__, both purposes are squeezed into vdso.
+    //   Since vdso will use randomized addresses (unlike the standard practice
+    //   for vsyscall), this implies that kernel calls on __x86__ can go through
+    //   randomized addresses, and so they need special treatment.
+    void *vdso = mtcp_sys_mmap(rinfo->vdsoStart,
+                                 vdsoEnd - vdsoStart,
+				 PROT_EXEC | PROT_WRITE | PROT_READ,
+                                 MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
```

Comment:
We need to also add MAP_FIXED or else the kernel might disregard our addr request.